File: backend/app/models/turno.py
from sqlalchemy import Date, ForeignKey, Integer, String, Time
from sqlalchemy.orm import relationship

from app.extensions import db


class Turno(db.Model):
    __tablename__ = "turnos"

    id_turno = db.Column(Integer, primary_key=True)
    fecha = db.Column(Date, nullable=False)
    hora_inicio = db.Column(Time, nullable=False)
    hora_fin = db.Column(Time, nullable=False)
    estado = db.Column(String(20), nullable=False)
    id_empleado = db.Column(Integer, ForeignKey("empleados.id_empleado"), nullable=False)
    id_puesto = db.Column(Integer, ForeignKey("puestos.id_puesto"), nullable=False)

    empleado = relationship("Empleado", back_populates="turnos", lazy="select")
    puesto = relationship("Puesto", back_populates="turnos", lazy="select")
    asistencias = relationship("Asistencia", back_populates="turno", lazy="select")
    novedades = relationship("Novedad", back_populates="turno", lazy="select")

    def __repr__(self) -> str:
        return f"<Turno {self.id_turno}>"

File: mobile/lib/features/shifts/models/turno.dart
class Turno {
  const Turno({
    this.idTurno,
    required this.fecha,
    required this.horaInicio,
    required this.horaFin,
    required this.estado,
    this.tipoTurno = '24 HORAS',
    this.tipoAsignacion = 'FIJO',
    required this.idEmpleado,
    required this.idPuesto,
    this.empleadoNombre,
    this.puestoNombre,
  });

  final int? idTurno;
  final String fecha;
  final String horaInicio;
  final String horaFin;
  final String estado;
  final String tipoTurno;
  final String tipoAsignacion;
  final int idEmpleado;
  final int idPuesto;
  final String? empleadoNombre;
  final String? puestoNombre;

  factory Turno.fromJson(Map<String, dynamic> json) {
    return Turno(
      idTurno: json['id_turno'] as int?,
      fecha: json['fecha'] as String,
      horaInicio: json['hora_inicio'] as String,
      horaFin: json['hora_fin'] as String,
      estado: json['estado'] as String,
      tipoTurno: json['tipo_turno'] as String? ?? '24 HORAS',
      tipoAsignacion: json['tipo_asignacion'] as String? ?? 'FIJO',
      idEmpleado: json['id_empleado'] as int,
      idPuesto: json['id_puesto'] as int,
      empleadoNombre:
          (json['empleado'] as Map<String, dynamic>?)?['nombre_completo']
              as String?,
      puestoNombre:
          (json['puesto'] as Map<String, dynamic>?)?['nombre_puesto']
              as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    'fecha': fecha,
    'hora_inicio': horaInicio,
    'hora_fin': horaFin,
    'estado': estado,
    'tipo_turno': tipoTurno,
    'tipo_asignacion': tipoAsignacion,
    'id_empleado': idEmpleado,
    'id_puesto': idPuesto,
  };
}
